File: app/modules/account_proxies/schemas.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import Field

from app.modules.shared.schemas import DashboardModel

ProxyStatus = Literal["untested", "testing", "working", "failed"]


class AccountProxyResponse(DashboardModel):
    id: str
    display_name: str
    redacted_proxy_url: str
    status: ProxyStatus
    last_tested_at: datetime | None = None
    last_test_error: str | None = None
    last_test_latency_ms: int | None = None
    created_at: datetime
    updated_at: datetime


class AccountProxiesResponse(DashboardModel):
    proxies: list[AccountProxyResponse] = Field(default_factory=list)


class AccountProxyUpsertRequest(DashboardModel):
    display_name: str = Field(min_length=1, max_length=255)
    proxy_url: str = Field(min_length=1)


class AccountProxyDraftTestRequest(DashboardModel):
    proxy_url: str = Field(min_length=1)


class AccountProxyTestResponse(DashboardModel):
    status: ProxyStatus
    last_tested_at: datetime
    last_test_error: str | None = None
    last_test_latency_ms: int | None = None


class AccountProxyDeleteResponse(DashboardModel):
    status: str

File: frontend/src/features/proxies/schemas.ts
import { z } from "zod";

export const AccountProxySchema = z.object({
  id: z.string(),
  displayName: z.string(),
  redactedProxyUrl: z.string(),
  status: z.enum(["untested", "testing", "working", "failed"]).or(z.string()),
  lastTestedAt: z.string().datetime({ offset: true }).nullable().optional(),
  lastTestError: z.string().nullable().optional(),
  lastTestLatencyMs: z.number().int().nullable().optional(),
  createdAt: z.string().datetime({ offset: true }),
  updatedAt: z.string().datetime({ offset: true }),
});

export const AccountProxiesResponseSchema = z.object({
  proxies: z.array(AccountProxySchema),
});

export const AccountProxyUpsertRequestSchema = z.object({
  displayName: z.string().min(1).max(255),
  proxyUrl: z.string().min(1),
});

export const AccountProxyDraftTestRequestSchema = z.object({
  proxyUrl: z.string().min(1),
});

export const AccountProxyTestResponseSchema = z.object({
  status: z.string(),
  lastTestedAt: z.string().datetime({ offset: true }),
  lastTestError: z.string().nullable().optional(),
  lastTestLatencyMs: z.number().int().nullable().optional(),
});

export const AccountProxyDeleteResponseSchema = z.object({
  status: z.string(),
});

export type AccountProxy = z.infer<typeof AccountProxySchema>;
export type AccountProxyUpsertRequest = z.infer<typeof AccountProxyUpsertRequestSchema>;
